src/time-picker/TimePicker.js

```diff
@@ -0,0 +1,211 @@
+import React from 'react';
+import { props, t, skinnable } from '../utils';
+import cx from 'classnames';
+import Dropdown from '../dropdown/Dropdown';
+import range from 'lodash/utility/range';
+import flatten from 'lodash/array/flatten';
+
+/*
+  Options is defined here cause we need to not change the ref.
+  This happend because of a very wired behaviour of react-select v0.6.12
+*/
+const options = [];
+
+const H24 = '24h';
+const H12 = '12h';
+const symbolRegex = /[\s,\.\|\/\\]/;
+const numberRegex = /^\d+$/;
+const time12Regex = /^([1-9]|1[0-2])[:,\.\|\/\\]?([0-5]\d)?\s*?(AM|PM)$/;
+const time24Regex = /^([01]?\d|2[0-3])[:,\.\|\/\\]?([0-5]?\d)?$/;
+const separator = ':';
+
+const pad = (num) => num <= 9 ? `0${num}` : num;
+const timeFormatter24 = (hour, minute) => `${pad(hour)}:${pad(minute)}`;
+const timeFormatter12 = (hour, minute) => {
+  if (hour === 0 || (hour > 12 && hour % 12 === 0)) {
+    return `12${separator}${pad(minute)} am`;
+  } else if (hour === 12) {
+    return `12${separator}${pad(minute)} pm`;
+  } else if (hour >= 13) {
+    return `${pad(hour % 12)}${separator}${pad(minute)} pm`;
+  } else {
+    return `${pad(hour)}${separator}${pad(minute)} am`;
+  }
+};
+
+const isValidTime = input => time12Regex.test(input) || time24Regex.test(input);
+
+const formatValue = (timeFormatter, hour, minute) => ({
+  value: `${hour}${separator}${minute}`,
+  label: timeFormatter(hour, minute)
+});
+
+const lteTime = (minTime, maxTime) => (
+  maxTime.hours > minTime.hours || (maxTime.hours === minTime.hours && maxTime.minutes >= minTime.minutes)
+);
+
+const addZero = str => {
+  if (str.length === 4) {
+    return `${str}0`;
+  } else {
+    return str;
+  }
+};
+
+const insertColon = str => {
+  if (numberRegex.test(str)) {
+    if (str.length === 3) {
+      const position = str.length - 1;
+      const strWithColon = `${str.substr(0, position)}${separator}${str.substr(position)}`;
+      return strWithColon;
+    } else if (str.length === 4) {
+      const position = str.length - 2;
+      const strWithColon = `${str.substr(0, position)}${separator}${str.substr(position)}`;
+      return strWithColon;
+    }
+  }
+  return str;
+};
+
+const cleanSeparator = str => str.replace(symbolRegex, separator);
+const cleanInput = inputStr => addZero(insertColon(cleanSeparator(inputStr)));
+
+const getSpecificOptionList = (cleanedInput, timeFormat) => {
```

Comment:
This is acutally the wirdest function here, because it will return the custom generated option list, that is different for 24h and 12h time formats. Of course 12h need to have an array of 2 options because we don't know since the very end (when he will click on one option) what he intends. This is actually a very custom behaviour for 12h. It works in a different way from what the formatter actually expect, because we don't have 2 different value (i.e. 11 & 23 => 11am & 11pm OR 0 & 12 => 12 am & 12pm ) but we have just the parsed value from the user input, not a generated one, and by design in can be always less or equal than 12, but as i said before we have to show the user both the choices. They looks similar but they are actually 2 completely different way of formatting. In order to use the already implemented we can use a simple solution, that is sum 12 to the input (that for sure is <= 12) this will return the exact opposite of the user input value, returning the results we need. But a very big issue here is that for managing this we have to expand the logic of the formatter for managing the special case of 24. This logic seems very complex, and actually it his, but i didnt have any other ideas that can smartly solve this problem without repeating code implementing 2 different formatter. Will be easier to understand reading the code.